File: Backend/dotnet/Services/GroupChatService.cs
using Microsoft.Extensions.Options;
using DotNetAgentFramework.Configuration;

namespace DotNetAgentFramework.Services;

/// <summary>
/// Group chat service implementation using Microsoft Agent Framework
/// </summary>
public class GroupChatService(
    IAgentService agentService,
    ISessionManager sessionManager,
    ILogger<GroupChatService> logger,
    IOptions<AzureAIConfig> azureConfig) : IGroupChatService
{
    private readonly IAgentService _agentService = agentService;
    private readonly ISessionManager _sessionManager = sessionManager;
    private readonly ILogger<GroupChatService> _logger = logger;
    private readonly AzureAIConfig _azureConfig = azureConfig.Value;

    /// <summary>
    /// Start a group chat with multiple agents using Microsoft Agent Framework
    /// </summary>
    public async Task<GroupChatResponse> StartGroupChatAsync(GroupChatRequest request)
    {
        using var cts = new CancellationTokenSource(TimeSpan.FromMinutes(4));
        var cancellationToken = cts.Token;
        
        var sessionId = request.SessionId ?? await _sessionManager.CreateSessionAsync();
        var messages = new List<GroupChatMessage>();
        var terminatedAgents = new HashSet<string>();

        // Add the initial user message
        var userMessage = new GroupChatMessage
        {
            Content = request.Message,
            Agent = "user",
            Timestamp = DateTime.UtcNow,
            Turn = 0,
            MessageId = Guid.NewGuid().ToString()
        };
        messages.Add(userMessage);
        await _sessionManager.AddMessageToSessionAsync(sessionId, userMessage);

        var currentTurn = 1;
        var startTime = DateTime.UtcNow;

        try
        {
            _logger.LogInformation("Starting Agent Framework group chat with {AgentCount} agents, max turns: {MaxTurns}", 
                request.Agents?.Count ?? 0, request.MaxTurns);

            // Process agents in sequence for each turn
            for (int turn = 1; turn <= request.MaxTurns; turn++)
            {
                cancellationToken.ThrowIfCancellationRequested();

                foreach (var agentName in request.Agents ?? new List<string>())
                {
                    cancellationToken.ThrowIfCancellationRequested();

                    // Skip agents that have terminated
                    if (terminatedAgents.Contains(agentName))
                    {
                        _logger.LogInformation("Agent {AgentName} has terminated, skipping", agentName);
                        continue;
                    }

                    var agentStartTime = DateTime.UtcNow;
                    var agent = await _agentService.GetAgentAsync(agentName);
                    if (agent == null)
                    {
                        _logger.LogWarning("Agent {AgentName} not found, skipping", agentName);
                        continue;
                    }

                    // Build context for the agent
                    var agentContext = BuildAgentContext(messages, agentName, request.Message);

                    try
                    {
                        // Get agent response with timeout
                        using var agentCts = new CancellationTokenSource(TimeSpan.FromMinutes(1));
                        using var combinedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, agentCts.Token);

                        var response = await agent.RespondAsync(request.Message, agentContext);

                        // Check if agent terminated
                        bool isTerminated = IsAgentTerminated(response);
                        if (isTerminated)
                        {
                            terminatedAgents.Add(agentName);
                            _logger.LogInformation("Agent {AgentName} terminated from conversation", agentName);
                        }

                        var agentMessage = new GroupChatMessage
                        {
                            Content = response,
                            Agent = agentName,
                            Timestamp = DateTime.UtcNow,
                            Turn = currentTurn,
                            MessageId = Guid.NewGuid().ToString(),
                            IsTerminated = isTerminated,
                            AgentType = "Agent Framework"
                        };

                        messages.Add(agentMessage);
                        await _sessionManager.AddMessageToSessionAsync(sessionId, agentMessage);

                        var processingTime = (DateTime.UtcNow - agentStartTime).TotalSeconds;
                        _logger.LogInformation("Agent {AgentName} responded in turn {Turn} ({ProcessingTime:F1}s), terminated: {IsTerminated}",
                            agentName, currentTurn, processingTime, isTerminated);
                        currentTurn++;
                    }
                    catch (OperationCanceledException)
                    {
                        _logger.LogWarning("Agent {AgentName} timed out, marking as terminated", agentName);
                        terminatedAgents.Add(agentName);

                        var timeoutMessage = new GroupChatMessage
                        {
                            Content = "TERMINATED - Agent response timed out",
                            Agent = agentName,
                            Timestamp = DateTime.UtcNow,
                            Turn = currentTurn,
                            MessageId = Guid.NewGuid().ToString(),
                            IsTerminated = true,
                            AgentType = "Agent Framework"
                        };
                        messages.Add(timeoutMessage);
                        currentTurn++;
                    }
                }

                // Check if all agents have terminated
                if (terminatedAgents.Count == request.Agents?.Count)
                {
                    _logger.LogInformation("All agents have terminated, ending group chat");
                    break;
                }
            }

            // Generate summary
            string? summary = null;
            try
            {
                using var summaryCts = new CancellationTokenSource(TimeSpan.FromSeconds(30));
                using var combinedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, summaryCts.Token);
                summary = await SummarizeConversationAsync(messages, combinedCts.Token);
            }
            catch (OperationCanceledException)
            {
                _logger.LogWarning("Conversation summary timed out, skipping");
                summary = "Summary generation timed out - conversation completed successfully.";
            }

            var totalProcessingTime = (int)(DateTime.UtcNow - startTime).TotalMilliseconds;

            return new GroupChatResponse
            {
                Messages = messages,
                SessionId = sessionId,
                TotalTurns = currentTurn - 1,
                Summary = summary,
                GroupChatType = "Agent Framework Group Chat",
                AgentCount = request.Agents?.Count ?? 0,
                TerminatedAgents = terminatedAgents.ToList(),
                TotalProcessingTimeMs = totalProcessingTime
            };
        }
        catch (OperationCanceledException)
        {
            _logger.LogWarning("Group chat cancelled due to timeout for session {SessionId}", sessionId);
            throw new TimeoutException("Group chat operation timed out. Please try with fewer agents or reduced max_turns.");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error during group chat for session {SessionId}", sessionId);
            throw;
        }
    }

    /// <summary>
    /// Summarize a conversation between multiple agents
    /// </summary>
    public async Task<string> SummarizeConversationAsync(List<GroupChatMessage> messages)
    {
        return await SummarizeConversationAsync(messages, CancellationToken.None);
    }

    /// <summary>
    /// Summarize a conversation with cancellation token support
    /// </summary>
    public async Task<string> SummarizeConversationAsync(List<GroupChatMessage> messages, CancellationToken cancellationToken = default)
    {
        try
        {
            if (messages.Count <= 1)
            {
                return "No meaningful conversation to summarize.";
            }

            var agentMessages = messages.Where(m => m.Agent != "user").ToList();
            if (!agentMessages.Any())
            {
                return "No agent responses to summarize.";
            }

            var userMessage = messages.FirstOrDefault(m => m.Agent == "user")?.Content ?? "No user message";
            
            // Create a comprehensive summary
            var agentNames = agentMessages.Select(m => m.Agent).Distinct().ToList();
            var totalMessages = agentMessages.Count;
            
            // Build a structured summary
            var summary = $"**Agent Framework Group Chat Summary**\n\n";
            summary += $"**Original Question**: {userMessage}\n\n";
            summary += $"**Participants**: {string.Join(", ", agentNames)} ({agentNames.Count} agents)\n";
            summary += $"**Total Responses**: {totalMessages}\n\n";
            
            summary += "**Key Contributions**:\n";
            foreach (var agentName in agentNames)
            {
                var agentMessagesForAgent = agentMessages.Where(m => m.Agent == agentName).ToList();
                var lastMessage = agentMessagesForAgent.LastOrDefault()?.Content;
                if (!string.IsNullOrEmpty(lastMessage))
                {
                    var preview = lastMessage.Length > 100 ? lastMessage.Substring(0, 100) + "..." : lastMessage;
                    summary += $"• **{agentName}**: {preview}\n";
                }
            }

            summary += $"\n**Conversation completed using Microsoft Agent Framework with {totalMessages} total responses from {agentNames.Count} agents.**";

            return summary;
        }
        catch (OperationCanceledException)
        {
            _logger.LogInformation("Summary generation was cancelled");
            return "Summary generation was cancelled due to timeout.";
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error generating conversation summary");
            return "Error generating summary - please review the conversation manually.";
        }
    }

    /// <summary>
    /// Check if an agent response indicates termination
    /// </summary>
    private static bool IsAgentTerminated(string response)
    {
        if (string.IsNullOrWhiteSpace(response))
            return false;

        return response.Trim().StartsWith("TERMINATED", StringComparison.OrdinalIgnoreCase);
    }

    /// <summary>
    /// Build context for an agent based on the conversation history
    /// </summary>
    private string BuildAgentContext(List<GroupChatMessage> currentMessages, string agentName, string userMessage)
    {
        var context = $"?? **User's Original Question**: {userMessage}\n\n";
        
        if (currentMessages.Count > 1)
        {
            var previousResponses = currentMessages
                .Where(m => m.Agent != "user" && m.Agent != agentName)
                .OrderBy(m => m.Turn)
                .Select(m => $"**{m.Agent}** (Turn {m.Turn}): {m.Content}")
                .ToList();

            if (previousResponses.Any())
            {
                context += $"?? **Other Agent Responses**:\n{string.Join("\n\n", previousResponses)}\n\n";
                context += $"?? **Your Role as {agentName}**:\n";
                context += "- Provide your unique perspective and expertise\n";
                context += "- Build upon or complement what others have said\n";
                context += "- Avoid repeating points already covered\n";
                context += "- Add new insights from your specialization\n";
                context += "- Be comprehensive yet concise\n\n";
                context += "Focus on what you can uniquely contribute to answer the user's question.";
            }
        }
        else
        {
            context += $"?? You are the first agent to respond. Provide a comprehensive answer from your {agentName} perspective.";
        }

        return context;
    }
}

File: Backend/python/services/group_chat_service.py
"""
Group chat service for multi-agent orchestration.

This service handles multi-agent conversations using the Microsoft Agent Framework,
similar to the .NET GroupChatService.
"""

import asyncio
import logging
from datetime import datetime
from typing import Dict, List, Optional, Any
from uuid import uuid4

from agent_framework import AgentThread, ChatMessage

from .agent_service import AgentService
from .session_manager import SessionManager
from models.chat_models import GroupChatRequest, GroupChatResponse, GroupChatMessage


logger = logging.getLogger(__name__)


class GroupChatService:
    """
    Service for managing multi-agent group conversations.
    
    This service provides functionality to:
    - Orchestrate conversations between multiple agents
    - Manage turn-based agent interactions
    - Handle agent thread creation and management
    - Format and respond with group chat results
    """
    
    def __init__(self, agent_service: AgentService, session_manager: SessionManager):
        """
        Initialize the group chat service.
        
        Args:
            agent_service: The agent service for managing agents
            session_manager: The session manager for conversation history
        """
        self.agent_service = agent_service
        self.session_manager = session_manager
        logger.info("GroupChatService initialized")
    
    async def start_group_chat(self, request: GroupChatRequest) -> GroupChatResponse:
        """
        Start a group chat with multiple agents.
        
        Args:
            request: The group chat request containing agents and message
            
        Returns:
            The group chat response with all agent interactions
        """
        try:
            session_id = request.session_id or await self.session_manager.create_session()
            messages = []
            terminated_agents = set()
            
            # Add the initial user message
            user_message = GroupChatMessage(
                content=request.message,
                agent="user",
                timestamp=datetime.utcnow().isoformat(),
                turn=0,
                message_id=str(uuid4())
            )
            messages.append(user_message)
            await self.session_manager.add_message_to_session(session_id, user_message)
            
            current_turn = 1
            start_time = datetime.utcnow()
            
            logger.info(f"Starting group chat with {len(request.agents)} agents, max turns: {request.max_turns}")
            
            # Process agents in sequence for each turn
            for turn in range(1, request.max_turns + 1):
                for agent_name in request.agents:
                    # Skip agents that have terminated
                    if agent_name in terminated_agents:
                        logger.info(f"Agent {agent_name} has terminated, skipping")
                        continue
                    
                    agent_start_time = datetime.utcnow()
                    
                    try:
                        # Get the agent
                        agent = await self.agent_service.get_agent(agent_name)
                        if not agent:
                            logger.warning(f"Agent {agent_name} not found, skipping")
                            continue
                        
                        # Build context for the agent
                        agent_context = self._build_agent_context(messages, agent_name, request.message)
                        
                        # Get agent response with timeout
                        try:
                            response_task = agent.run(request.message)
                            response = await asyncio.wait_for(response_task, timeout=60.0)  # 1 minute timeout
                            
                            # Extract response content
                            response_content = response.text if hasattr(response, 'text') else str(response)
                            
                            # Check if agent terminated
                            is_terminated = self._is_agent_terminated(response_content)
                            if is_terminated:
                                terminated_agents.add(agent_name)
                                logger.info(f"Agent {agent_name} terminated from conversation")
                            
                            agent_message = GroupChatMessage(
                                content=response_content,
                                agent=agent_name,
                                timestamp=datetime.utcnow().isoformat(),
                                turn=current_turn,
                                message_id=str(uuid4()),
                                metadata={
                                    "agent_type": getattr(agent, 'agent_type', agent_name),  # Fallback to agent_name if agent_type missing
                                    "processing_time": (datetime.utcnow() - agent_start_time).total_seconds(),
                                    "terminated": is_terminated
                                }
                            )
                            
                            messages.append(agent_message)
                            await self.session_manager.add_message_to_session(session_id, agent_message)
                            
                            logger.debug(f"Agent {agent_name} responded in turn {current_turn}")
                            
                        except asyncio.TimeoutError:
                            logger.error(f"Agent {agent_name} timed out in turn {current_turn}")
                            # Create timeout message
                            timeout_message = GroupChatMessage(
                                content=f"Agent {agent_name} did not respond within the timeout period.",
                                agent=agent_name,
                                timestamp=datetime.utcnow(),
                                turn=current_turn,
                                message_id=str(uuid4()),
                                metadata={
                                    "error": "timeout",
                                    "processing_time": 60.0
                                }
                            )
                            messages.append(timeout_message)
                            terminated_agents.add(agent_name)
                    
                    except Exception as e:
                        logger.error(f"Error with agent {agent_name} in turn {current_turn}: {str(e)}")
                        # Create error message
                        error_message = GroupChatMessage(
                            content=f"Error occurred with agent {agent_name}: {str(e)}",
                            agent=agent_name,
                            timestamp=datetime.utcnow().isoformat(),
                            turn=current_turn,
                            message_id=str(uuid4()),
                            metadata={
                                "error": str(e),
                                "processing_time": (datetime.utcnow() - agent_start_time).total_seconds()
                            }
                        )
                        messages.append(error_message)
                        terminated_agents.add(agent_name)
                
                current_turn += 1
                
                # Check if all agents have terminated
                if len(terminated_agents) >= len(request.agents):
                    logger.info("All agents have terminated, ending group chat")
                    break
            
            total_time = (datetime.utcnow() - start_time).total_seconds()
            
            # Create group chat response
            response = GroupChatResponse(
                messages=messages,
                session_id=session_id,
                total_turns=current_turn - 1,
                participating_agents=request.agents,
                terminated_agents=list(terminated_agents),
                total_processing_time=total_time,
                metadata={
                    "max_turns": request.max_turns,
                    "format": request.format,
                    "start_time": start_time.isoformat(),
                    "end_time": datetime.utcnow().isoformat()
                }
            )
            
            logger.info(f"Completed group chat with {len(messages)} messages in {total_time:.2f} seconds")
            return response
            
        except Exception as e:
            logger.error(f"Failed to start group chat: {str(e)}")
            raise
    
    def _build_agent_context(self, messages: List[GroupChatMessage], agent_name: str, original_message: str) -> str:
        """
        Build context for an agent based on conversation history.
        
        Args:
            messages: List of messages in the conversation
            agent_name: Name of the current agent
            original_message: The original user message
            
        Returns:
            Context string for the agent
        """
        try:
            context_parts = [f"Original user message: {original_message}"]
            
            # Add recent conversation history (last 5 messages)
            recent_messages = messages[-5:] if len(messages) > 5 else messages
            
            if len(recent_messages) > 1:  # More than just the user message
                context_parts.append("Recent conversation:")
                for message in recent_messages[1:]:  # Skip the first user message
                    if message.agent != agent_name:  # Don't include own messages
                        context_parts.append(f"- {message.agent}: {message.content[:200]}...")
            
            context = "\n".join(context_parts)
            return context
            
        except Exception as e:
            logger.error(f"Failed to build agent context: {str(e)}")
            return original_message
    
    def _is_agent_terminated(self, response_content: str) -> bool:
        """
        Check if an agent has indicated termination.
        
        Args:
            response_content: The agent's response content
            
        Returns:
            True if the agent has terminated, False otherwise
        """
        try:
            # Look for termination indicators in the response
            termination_indicators = [
                "TERMINATE",
                "CONVERSATION_COMPLETE",
                "END_CONVERSATION",
                "I'm done",
                "conversation is complete",
                "no further assistance needed"
            ]
            
            response_lower = response_content.lower()
            return any(indicator.lower() in response_lower for indicator in termination_indicators)
            
        except Exception as e:
            logger.error(f"Error checking agent termination: {str(e)}")
            return False
    
    async def get_group_chat_templates(self) -> List[Dict[str, Any]]:
        """
        Get available group chat templates.
        
        Returns:
            List of available templates
        """
        try:
            # Define comprehensive templates matching .NET version
            templates = [
                {
                    "name": "general_inquiry",
                    "display_name": "General Inquiry",
                    "description": "General multi-agent inquiry with people lookup and knowledge search",
                    "agents": ["people_lookup", "knowledge_finder"],
                    "max_turns": 3,
                    "format": "user_friendly",
                    "use_cases": ["Employee questions", "Information lookup", "General assistance"],
                    "config": {
                        "response_synthesis": True,
                        "turn_management": "sequential",
                        "timeout_per_agent": 60
                    }
                },
                {
                    "name": "comprehensive_research",
                    "display_name": "Comprehensive Research",
                    "description": "Comprehensive research using all available agents",
                    "agents": ["generic_agent", "people_lookup", "knowledge_finder"],
                    "max_turns": 2,
                    "format": "detailed",
                    "use_cases": ["Complex research", "Multi-perspective analysis", "Detailed investigations"],
                    "config": {
                        "response_synthesis": False,
                        "turn_management": "sequential",
                        "timeout_per_agent": 90
                    }
                },
                {
                    "name": "people_focused",
                    "display_name": "People-Focused Inquiry",
                    "description": "People-focused inquiry with general assistance",
                    "agents": ["people_lookup", "generic_agent"],
                    "max_turns": 2,
                    "format": "user_friendly",
                    "use_cases": ["HR inquiries", "Team information", "Contact lookup"],
                    "config": {
                        "response_synthesis": True,
                        "turn_management": "sequential",
                        "timeout_per_agent": 45
                    }
                },
                {
                    "name": "knowledge_deep_dive",
                    "display_name": "Knowledge Deep Dive",
                    "description": "Deep knowledge search with expert analysis",
                    "agents": ["knowledge_finder", "generic_agent"],
                    "max_turns": 3,
                    "format": "user_friendly",
                    "use_cases": ["Policy questions", "Procedure lookup", "Documentation search"],
                    "config": {
                        "response_synthesis": True,
                        "turn_management": "sequential",  
                        "timeout_per_agent": 75
                    }
                }
            ]
            
            return templates
            
        except Exception as e:
            logger.error(f"Failed to get group chat templates: {str(e)}")
            return []
    
    async def create_from_template(self, template_name: str, message: str, session_id: Optional[str] = None) -> GroupChatResponse:
        """
        Create a group chat from a template.
        
        Args:
            template_name: Name of the template to use
            message: The user message
            session_id: Optional session ID
            
        Returns:
            The group chat response
        """
        try:
            templates = await self.get_group_chat_templates()
            template = next((t for t in templates if t["name"] == template_name), None)
            
            if not template:
                raise ValueError(f"Template {template_name} not found")
            
            request = GroupChatRequest(
                message=message,
                agents=template["agents"],
                session_id=session_id,
                max_turns=template["max_turns"],
                format=template["format"]
            )
            
            return await self.start_group_chat(request)
            
        except Exception as e:
            logger.error(f"Failed to create group chat from template {template_name}: {str(e)}")
            raise